src/grammar/udt.rs

```diff
@@ -0,0 +1,1052 @@
+            parse_datatype(p);
+            if !p.eat(T![,]) || p.at(T![")"]) {
+                break;
+            }
+            if [
+                T![constructor],
+                T![final],
+                T![instantiable],
+                T![map],
+                T![member],
+                T![not],
+                T![order],
+                T![overriding],
+                T![static],
+            ]
+            .contains(&p.current())
+            {
+                break;
+            }
+        });
+        if [
+            T![constructor],
+            T![final],
+            T![instantiable],
+            T![map],
+            T![member],
+            T![not],
+            T![order],
+            T![overriding],
+            T![static],
+        ]
+        .contains(&p.current())
+        {
+            safe_loop!(p, {
+                parse_element_spec(p);
+                if !p.at(T![,]) || p.at(T![")"]) {
+                    break;
+                }
+            });
+        }
+        p.expect(T![")"]);
+    }
+    safe_loop!(p, {
+        let mut ate_something = false;
+        ate_something |= p.eat(T![not]);
+        ate_something |= p.eat_one_of(&[T![final], T![instantiable], T![persistable]]);
+        if p.at(T![;]) || !ate_something {
+            break;
+        }
+    });
+
+    p.finish();
+}
+
+fn parse_varray_type_spec(p: &mut Parser) {
+    p.start(SyntaxKind::VarrayTypeSpec);
+    if p.eat(T![varying]) {
+        p.expect(T![array]);
+    } else {
+        p.expect(T![varray]);
+    }
+    p.expect(T!["("]);
+    p.expect(T![int_literal]);
+    p.expect(T![")"]);
+    p.expect(T![of]);
+    let expect_rparen = p.eat(T!["("]);
+    parse_datatype(p);
+    if p.eat(T![not]) {
+        p.expect(T![null]);
+    }
+    if expect_rparen {
+        p.expect(T![")"]);
+    }
+    p.eat(T![not]);
+    p.eat(T![persistable]);
+    p.finish();
+}
+
+fn parse_nested_table_type_spec(p: &mut Parser) {
+    p.start(SyntaxKind::NestedTableTypeSpec);
+    p.expect(T![table]);
+    p.expect(T![of]);
+    let expect_rparen = p.eat(T!["("]);
+    parse_datatype(p);
+    if p.eat(T![not]) {
+        p.expect(T![null]);
+    }
+    if expect_rparen {
+        p.expect(T![")"]);
+    }
+    p.eat(T![not]);
+    p.eat(T![persistable]);
+    p.finish();
+}
+
+#[cfg(test)]
+mod tests {
+    use expect_test::expect;
+
+    use crate::grammar::tests::{check, parse};
+
+    use super::parse_udt;
+
+    #[test]
+    fn simple_udt() {
+        check(
+            parse(
+                "CREATE TYPE customer_typ_demo AS OBJECT ( customer_id NUMBER(6)
+    , cust_first_name    VARCHAR2(20)
+    , cust_last_name     VARCHAR2(20)
+    , cust_address       CUST_ADDRESS_TYP
+    , phone_numbers      PHONE_LIST_TYP
+    , nls_language       VARCHAR2(3)
+    , nls_territory      VARCHAR2(30)
+    , credit_limit       NUMBER(9,2)
+    , cust_email         VARCHAR2(30)
+    , cust_orders        ORDER_LIST_TYP
+    ) ;",
```

Comment:
Lets reduce this to the relevant parts:
```suggestion
                "CREATE TYPE customer_typ_demo AS OBJECT ( customer_id NUMBER(6)
    , cust_address       CUST_ADDRESS_TYP
    ) ;",
```